### ISSUE TITLE
SPAs and Unit Quirks Disappeared

### ISSUE THREAD

**Issue by scJazz2** [MegaMek/megamek#21]
MM Version 0.41.11

During this game I noticed that my pilot SPAs and Unit Quirks disappeared.

At first I thought it was merely a problem with showing the information on the Unit Display window. But when I saved the MUL at the end of the game all the SPAs and Unit Quirks were missing.

Attached: 
Devil Dogs Display Bug Before.mul - the MUL used to start the game
Devil Dogs Display Bug After.mul - the MUL saved at the end
megameklog.txt - no errors
display_bug save game
epic (2).zip - Customs for the game
Disappearing SPAs.jpg - a screen shot showing a unit with Jumping Jack from the MUL but missing Jumping Jack in game.

[SPA Issue.zip](https://github.com/MegaMek/megamek/files/120349/SPA.Issue.zip)

**Comment by arlith** [comment]
I see your screenshot, and I see your exported MUL.  But, the saved game you sent still has the SPAs.  It would help to have more information on when they disappeared.  If you continue the game from that save will it happen?  Did it happen after reloading a save?  I glanced through the code and I didn't find any areas that seemed like they would clear SPAs.  Without having more information about how this happened, it's going to be tough to hunt this down.

**Comment by pheonixstorm** [comment]
Perhaps mm is not loading the spa options properly from the save. Only idea I have if its in the save itself

**Comment by scJazz2** [comment]
OK to add some narrative to this issue. The game was played on a dedicated server. The game was played in 3 sessions. The computer running the ded was not rebooted over the course of the week from start to game finish although it was slept.

The SaveGame I submitted was made during the same session as the screenshot. So this is a "runtime" error. The dedicated server is still running and I just issued /load from the exact point I made the screenshot. SPAs are not appearing. This is after victory/defeat and MM returning to the Lobby. What ever is broken is still running on the ded.

While the ded was running I crashed MHQ/MM at least twice and had to END TASK on these instances. The ded was never touched.

A few months ago it was noted in our Campaign that using Double Blind meant that we couldn't see the opponents SPA/Quirks. So I didn't really think much of it. Now I think, I was seeing this same bug in action.

Arlith, since you joined as an OPFOR in this campaign you know my email address. The dedicated server is still running and the broken state is currently loaded. If you want to connect email me or hit me up on YIM.

**Comment by scJazz2** [comment]
Ohhh and somewhat similar strange has been seen when I run ASF battles. 2 sessions on the same computer causing the buttons on the main screen to change.

**Comment by arlith** [comment]
I would be somewhat surprised if loading the game is the issue, as that's basically taken care of by a third-party library (libxstream).  My hunch with this is that somehow the option gets turned off, but I don't know how that would happen.  The saved game attached has the option on and teh SPAs saved, as I can see  them when I load the game.

**Comment by scJazz2** [comment]
Could it somehow be related to the fact that MM launched from MHQ froze a couple of times and had to be Ended while the Ded was running?

**Comment by arlith** [comment]
Probably not.  Did you check the logs?  It shouldn't matter if MM is launched from MHQ or not, but if the server freezes, you should check the logfile for an uncaught exception.  The lack of an exception is significant....

**Comment by scJazz2** [comment]
Maybe I am not being clear...
Playing AtB/MM sessions launched from MHQ
While the Ded was running
Had to kill the AtB/MM session and MHQ a few times while the Dedicated was running.

Did Java throw away something useful? I mention this only because I've seen an intersection of 2 MM sessions running on the same computer causing changes to the Buttons.

There was nothing in the logs.

**Comment by arlith** [comment]
Ah.  So, the client state can definitely change.  There are certain things (like button order) that is purely important to the client.  The server knows nothing about a UI.  All of the game state is encapsulated within the Game object (which is what is serialized into a saved game).  You should not lose any important state by having the client leave (and potentially restart) multiple times within a game.  If that happened there would be an issue, but I don't think it's the case here.

**Comment by scJazz2** [comment]
I accept that you don't think it is the issue. However, I am nearly certain that it is the problem.

**Comment by ObviousTech** [comment]
Here is a game where this happened.  The autosave is from the start of the movement phase.  The other save is a manual save after some of the units have moved.

When loading the mid-phase save, all SPAs and quirks are gone.

There are no errors in the mekhq logs.

[autosave_2016-05-08_02-17-49.sav.gz](https://github.com/MegaMek/megamek/files/254196/autosave_2016-05-08_02-17-49.sav.gz)
[grandbase.sav.gz](https://github.com/MegaMek/megamek/files/254197/grandbase.sav.gz)

**Comment by scJazz2** [comment]
More info... as I just saw it in another game. The game is double blind. Player 1 can see all the Quirks and SPAs, Player 1 can not. The Quirks and SPAs are actually active but not showing in Unit Display.

**Comment by arlith** [comment]
I came across this testing other issues, and I figured it out.  Both `Quirks` and `PilotOptions` use the `AbstractOptions` framework in MM.  For this explanation, I will only refer to `PilotOptions` but everything I say also holds for `Quirks`.  Essentially, each of these "options" has a singleton (static instance) implementation of `AbstractOptionsInfo`.  This basically keeps track of the "option groups" and options themselves.  This information gets initialized in `PilotOptions.initialize()`.  The problem is that, `PilotOptions.initialize()` is only ever called when a `PilotOptions` instance is instantiated.  When a saved game is loaded, this never happens.  If `PilotOptions.initialize()` isn't called, then `PilotOptionsInfo` doesn't know about the groups and options, so when the UI code uses it for display purposes, nothing is displayed.

To fix this, I updated `PilotOptionsInfo` to have a static `initialized` state, and if it hasn't been initialized, it creates a dumby `PilotOptions` to force `PilotOptions.initialize()` to be called.  This seems somewhat kludgey, but is an easy fix without having to duplicate code.
